modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/topology/PartitionsEvictManager.java

```diff
@@ -791,10 +951,312 @@ private int calculateBucket() {
+            }
+
+            scheduleNextTask(0);
+        }
+
+        /**
+         * Cancels index purge tasks.
+         */
+        public void stop() {
+            if (log.isInfoEnabled())
+                log.info("Exclusive purge : stop");
+
+            resFut.onDone(new IgniteCheckedException("Cache is stopping."));
+
+            try {
+                cancelFut.cancel();
+            }
+            catch (IgniteCheckedException e) {
+                log.error("Error cancelling index purge tasks", e);
+            }
+        }
+    }
+
+    /**
+     * Purges partitions from single index.
+     */
+    private class IndexPurgeTask extends AbstractEvictionTask {
+        /** Actual worker that removes rows from the index. */
+        Runnable runnable;
+
+        /** Future internal to the worker. */
+        IgniteInternalFuture<?> fut;
+
+        /**
+         *
+         * @param grpEvictionCtx Group eviction context.
+         * @param runnable Actual worker that removes rows from the index.
+         * @param fut Future internal to the worker.
+         */
+        IndexPurgeTask(GroupEvictionContext grpEvictionCtx, Runnable runnable, IgniteInternalFuture<?> fut) {
+            super(null, grpEvictionCtx, TaskType.INDEX_PURGE);
+
+            this.runnable = runnable;
+
+            this.fut = fut;
+
+            fut.listen(f -> {
+                if (f.error() == null)
+                    finishFut.onDone();
+                else
+                    finishFut.onDone(f.error());
+            });
+        }
+
+        /** {@inheritDoc} */
+        @Override boolean run0() {
+            runnable.run();
+
+            return true;
+        }
+
+        /** {@inheritDoc} */
+        @Override void scheduleRetry() {
+            // No-op.
+        }
+    }
+
+    /**
+     * Removes links belonging to certain partitions from H2RowCache.
+     */
+    private class RowCacheCleanerTask extends AbstractEvictionTask {
+        /** Row cache cleaner. */
+        GridQueryRowCacheCleaner cleaner;
+
+        /** Partitions. */
+        Set<Integer> parts;
+
+        /**
+         *
+         * @param grpEvictionCtx Group eviction context.
+         * @param cleaner Row cache cleaner.
+         * @param parts Partitions.
+         */
+        RowCacheCleanerTask(GroupEvictionContext grpEvictionCtx, GridQueryRowCacheCleaner cleaner, Set<Integer> parts) {
+            super(null, grpEvictionCtx, TaskType.CLEAR_ROWCACHE);
+
+            this.cleaner = cleaner;
+            this.parts = parts;
+        }
+
+        /** {@inheritDoc} */
+        @Override boolean run0() {
+            cleaner.removeByPartition(parts);
+
+            return true;
+        }
+
+        /** {@inheritDoc} */
+        @Override void scheduleRetry() {
+            // No-op.
+        }
+    }
+
+    /**
+     * Performs per-partition actions in scope of exclusive partition set purge.
+     */
+    private class ClearOnheapEntriesTask extends AbstractEvictionTask {
```

Comment:
Le'ts use consistent naming and rename it to `PurgeOnheapEntriesTask` or something similar.